### ISSUE TITLE
Leftover .ldb file

### ISSUE THREAD

**Issue by pwnall** [google/leveldb#593]
While validating [the design for a feature on top of leveldb](https://docs.google.com/document/d/16_igCI15Gfzb6UYqeuJTmJPrzEtawz6Y1tVOKNtYgiU/), we've been stress-testing a database using a workload consisting of Put() and Delete(). We consistently see a .ldb file being left behind, so it seems like compaction doesn't delete all the old files.

This may be related to #573, which also suggests a potential accounting problem in compaction.

**Comment by dmurph** [comment]
Yep. More specifically, I'm putting batches of data, while keeping some alive temporarily, before deleting them:

1. Put batch 1
2. Put batch 2
3. Put batch 3, delete batch 1
4. Put batch 4, delete batch 2,
5. etc.

The first level file should be full of data that is soon deleted, but it sticks around forever.

**Comment by pedrinimm** [comment]
Hi Guys, 

Any update on this issue, I am facing the same issue.
In my case I have a stream of incoming data which is converted into a struct, saved while processing then removed.

Same as @dmurph.  But after a while, my amount of ldb files just grows and grows. All not used.
Because, I wrote a small binary that does only CompactRange from Start to Finish and all get removed.

**Comment by pwnall** [comment]
@dmurph Can you please dust off your experiment's code and answer @cmumford's question above?

**Comment by dmurph** [comment]
@pedrinimm That might be a different issue - can you give a range of 'growth' that you see? Or is it just the 00005.ldb file that sticks around?

@cmumford I will investigate, I'm not sure.

**Comment by dmurph** [comment]
@cmumford Yes - I just re-ran my test with a persistent directory instead of a new one, and it keeps the 000005.ldb file around even after the second run of the test.

**Comment by dmurph** [comment]
Updated the test to use a persistent folder: https://chromium-review.googlesource.com/c/chromium/src/+/1080339

**Comment by nysenthil** [comment]
In my team, we are seeing the same issue in one of our large enterprise customer projects. Old .ldb files are being left without getting deleted. It eventually leads to filling up the disk after two days. Is there a fix or workaround or any advice from the Level DB team to avoid this critical problem?

Thank you.

**Comment by cmumford** [comment]
@nysenthil check that your code doesn't have any unreleased snapshots or iterators. Also check the log (*.log) that your Env may be creating for evidence of problems. This bug is only for a single *.ldb file being left behind - I'm unaware of a bug that fails to delete many *.ldb files.

**Comment by nysenthil** [comment]
Hi Chris,
The problem I described above is much more related to this open issue by not purging the old .ldb files: Disk space used by the DB grows without bounds #603
